### PR TITLE
feat: pretty-print for iterative deployment

### PR DESCRIPTION
**Pull request by jhockett**
#### Description of changes
This is a continuation of https://github.com/aws-amplify/amplify-cli/pull/10909
> This change introduces pretty-print log feature into iterative deployment.

I rebased and then addressed some lint issues. The majority of the work was done by @sreekanth-sreekumar.

#### Description of how you validated changes
Both unit tests and e2e tests passed. Manually tested adding and removing multiple GSI's manually through schema changes as well as `amplify api rebuild`.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-provider-awscloudformation/src/__tests__/iterative-deployment/stack-progress-printer.test.ts [hunk 1]
```diff
@@ -2,7 +2,16 @@ import { StackProgressPrinter } from '../../iterative-deployment/stack-progress-
...
+  projectName: 'test',
+  envName: 'dev',
+  rootStackName: 'root-app',
+  rootResources: [{ key: 'test-app', category: 'api' }],
+  categories: [{ name: 'api', size: 2 }],
+  eventToCategories: new Map(),
+};
+
+const printer: IStackProgressPrinter = new StackProgressPrinter(eventMap);
 jest.mock('columnify');
```
Comment (awsluja) [inline packages/amplify-provider-awscloudformation/src/__tests__/iterative-deployment/stack-progress-printer.test.ts:16]: **columnify** is not being called when testing locally due to TTY detection;
causing this test to fail:
image

Comment (awsluja) [reply packages/amplify-provider-awscloudformation/src/__tests__/iterative-deployment/stack-progress-printer.test.ts:16]: Related to this , stack-progres-printer.ts

image

**Review by awsluja — CHANGES_REQUESTED** [review]

File: packages/amplify-provider-awscloudformation/src/__tests__/iterative-deployment/stack-progress-printer.test.ts (hunk 1 above)
Comment (awsluja) [reply packages/amplify-provider-awscloudformation/src/__tests__/iterative-deployment/stack-progress-printer.test.ts:16]: After a bit more review; this test is only for the "defaultLogs".
A new set of tests need to be designed & added for the pretty printer.

File: packages/amplify-provider-awscloudformation/src/__tests__/iterative-deployment/stack-progress-printer.test.ts [hunk 2]
```diff
@@ -20,8 +18,51 @@ jest.mock('chalk', () => ({
...
+      Timestamp: new Date('2021-01-01'),
+    });
+    printer.addActivity({
+      StackId: 'test',
+      EventId: 'test',
+      StackName: 'test',
+      Timestamp: new Date('2020-01-01'),
+    });
+    printer.print();
+    expect(columnify).not.toBeCalled();
```
Comment (awsluja) [inline packages/amplify-provider-awscloudformation/src/__tests__/iterative-deployment/stack-progress-printer.test.ts:59]: is there anything we can test about the console output for pretty print?

Comment (awsluja) [reply packages/amplify-provider-awscloudformation/src/__tests__/iterative-deployment/stack-progress-printer.test.ts:59]: this is covered by other tests (progressbar/multibar tests); resolving

File: packages/amplify-prompts/src/progressbars/spinner.ts [hunk 3]
```diff
@@ -3,64 +3,83 @@
...
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+    const lines = [{
+      renderString: `${this.frames[this.frameCount]} ${this.prefixText}`,
+      color: '',
+    }];
+    this.frameCount = ++this.frameCount % this.frames.length;
+    if (this.terminal) {
       this.terminal.writeLines(lines);
```
Comment (sobolk) [inline packages/amplify-prompts/src/progressbars/spinner.ts:40]: does it make sense to execute other logic in this method / class if there's no terminal ?
Should this whole thing just be no-op (aka NullObject) if there's no terminal?

File: packages/amplify-prompts/src/progressbars/spinner.ts [hunk 4]
```diff
@@ -3,64 +3,83 @@
...
-      this.prefixText = this.prefixText.replace('\n', '');
-      this.terminal.cursor(false);
-      this.render();
+  /**
+   * Starts a spinner and calls render function.
+   */
+  start(text: string | null): void {
+    if (!this.terminal) {
+      this.terminal = new AmplifyTerminal();
     }
```
Comment (sobolk) [inline packages/amplify-prompts/src/progressbars/spinner.ts:51]: what's the reason to move it from ctor here? it seems to be introducing complexity to this class.

File: packages/amplify-prompts/src/progressbars/spinner.ts [hunk 5]
```diff
@@ -3,64 +3,83 @@
...
+    const lines : StringObj[] = [{
+      renderString: text || '',
+      color: success ? 'green' : 'red',
+    }];
+
+    clearTimeout(this.timer);
+    this.terminal.writeLines(lines);
+    this.terminal.cursor(true);
+    this.terminal = null;
```
Comment (sobolk) [inline packages/amplify-prompts/src/progressbars/spinner.ts:83]: should a spinner be single usage? Does TS have a concept of disposable (aka closeable) ?
how do we prevent bad usage? I.e. what happens if multiple async stacks share reference to this?

File: packages/amplify-provider-awscloudformation/src/aws-utils/aws-cfn-progress-formatter.ts [hunk 6]
```diff
@@ -1,85 +1,16 @@
...
+  CFN_SUCCESS_STATUS,
+  CNF_ERROR_STATUS,
+  EventMap,
+} from '../utils/progress-bar-helpers';
 
 /**
  * Initializing the root and individual category bars
  */
-const initializeProgressBars = (eventMap : EventMap) : MultiProgressBar => {
+export const initializeProgressBars = (eventMap: EventMap): MultiProgressBar => {
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/aws-utils/aws-cfn-progress-formatter.ts:13]: does this become public api of this package?

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 7]
```diff
@@ -46,39 +56,46 @@ export class DeploymentError extends Error {
...
 /**
- *
+ * Deployment step type
  */
 export type DeploymentStep = {
   deployment: DeploymentOp;
   rollback: DeploymentOp;
 };
 
+type EventMap = {
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:73]: this doesn't look like map. what is this type representing? what kind of event is this?

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 8]
```diff
@@ -46,39 +56,46 @@ export class DeploymentError extends Error {
...
 export class DeploymentManager {
   /**
    * Helper method to get an instance of the Deployment manager with the right credentials
    */
-
   public static createInstance = async (
     context: $TSContext,
     deploymentBucket: string,
-    spinner: ora.Ora,
+    eventMap: EventMap,
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:92]: do export this function from this package?

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 9]
```diff
@@ -154,19 +174,28 @@ export class DeploymentManager {
...
-            if (state.matches('idle')) {
-              this.logSpinnerMessage('deploy', `Starting deployment`);
-            } else if (state.matches('deploy.waitForTablesToBeReady')) {
-              this.logSpinnerMessage('deploy', `Waiting for DynamoDB indices to be ready`);
-            } else if (state.matches('deploy')) {
-              this.logSpinnerMessage('deploy', `Deploying (${maxDeployed} of ${state.context.stacks.length})`);
-            } else if (state.matches('rollback')) {
-              this.logSpinnerMessage('deploy', `Rolling back (${maxDeployed - state.context.currentIndex} of ${maxDeployed})`);
+
+            const key = Object.keys(deploySpinnerMessage).find(elm => state.matches(elm));
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:179]: nit: elm?

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 10]
```diff
@@ -1,37 +1,47 @@
...
+const deploySpinnerMessage = {
+  idle: { machine: 'deploy', message: `Starting deployment.` },
+  'deploy.waitForTablesToBeReady': { machine: 'deploy', message: `Waiting for DynamoDB indices to be ready.` },
+};
+
+const rollbackSpinnerMessage = {
+  idle: { machine: 'rollback', message: `Starting rollback.` },
+  preRollback: { machine: 'rollback', message: `Waiting for previous deployment to finish.` },
+  'rollback.waitForTablesToBeReady': { machine: 'rollback', message: `Waiting for DynamoDB indices to be ready.` },
+};
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:40]:
```suggestion
const deploySpinnerMessages = {
  idle: { machine: 'deploy', message: `Starting deployment.` },
  'deploy.waitForTablesToBeReady': { machine: 'deploy', message: `Waiting for DynamoDB indices to be ready.` },
};

const rollbackSpinnerMessages = {
  idle: { machine: 'rollback', message: `Starting rollback.` },
  preRollback: { machine: 'rollback', message: `Waiting for previous deployment to finish.` },
  'rollback.waitForTablesToBeReady': { machine: 'rollback', message: `Waiting for DynamoDB indices to be ready.` },
};
```

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts (hunk 9 above)
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:179]: `deploySpinnerMessage` not being plural makes this piece hard to parse on first attempt.

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 11]
```diff
@@ -154,19 +174,28 @@ export class DeploymentManager {
...
-            } else if (state.matches('deploy.waitForTablesToBeReady')) {
-              this.logSpinnerMessage('deploy', `Waiting for DynamoDB indices to be ready`);
-            } else if (state.matches('deploy')) {
-              this.logSpinnerMessage('deploy', `Deploying (${maxDeployed} of ${state.context.stacks.length})`);
-            } else if (state.matches('rollback')) {
-              this.logSpinnerMessage('deploy', `Rolling back (${maxDeployed - state.context.currentIndex} of ${maxDeployed})`);
+
+            const key = Object.keys(deploySpinnerMessage).find(elm => state.matches(elm));
+            if (key) {
+              const keyObj = deploySpinnerMessage[key];
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:181]: what's key what's keyObj ? can we name it after what they reference to ?

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 12]
```diff
@@ -175,6 +204,8 @@ export class DeploymentManager {
               return resolve();
             case 'rolledBack':
             case 'failed':
+              this.printer.stopBars();
+              promptsPrinter.info(`Rolled back (${maxDeployed - state.context.currentIndex} of ${maxDeployed})`);
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:208]: probably for future. Can we replace this massive `switch-case` with event handlers ?

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 13]
```diff
@@ -220,25 +251,29 @@ export class DeploymentManager {
...
-              this.logSpinnerMessage('rollback', `Waiting for previous deployment to finish`);
-            } else if (state.matches('rollback.waitForTablesToBeReady')) {
-              this.logSpinnerMessage('rollback', `Waiting for DynamoDB indices to be ready`);
-            } else if (state.matches('rollback')) {
-              this.logSpinnerMessage('rollback', `Rolling back (${maxDeployed - state.context.currentIndex} of ${maxDeployed})`);
+            const key = Object.keys(rollbackSpinnerMessage).find(elm => state.matches(elm));
+            if (key) {
+              const keyObj = rollbackSpinnerMessage[key];
+              this.logger(keyObj.machine, [{ spinner: keyObj.message }])();
+              this.spinner.start(keyObj.message);
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:258]: this is duplicated code. 
Could we add top level routing to this logic that pick's message map first?
Or change key convention and have just one map?

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 14]
```diff
@@ -220,25 +251,29 @@ export class DeploymentManager {
...
             } else if (state.matches('rolledBack')) {
-              this.spinner.succeed(`Rolled back succesfully`);
-              this.logger('rollback', [{ spinner: 'Rolled back successfully' }])();
+              this.spinner.stop();
+              this.printer.finishBars();
+              this.printer.stopBars();
+              promptsPrinter.info(`Rolled back (${maxDeployed - state.context.currentIndex} of ${maxDeployed})`);
+              this.logger('rollback', [{ spinner: 'Rolled back successfully' }]);
+            } else if (state.matches('rollback')) {
+              this.spinner.stop();
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:266]: should stop implicitly stop bars ? we seem to do a lot of low level operations for something that should look like.
```
spinner = new Spiner()
try {
spinner.start()
} finally {
spinner stop():
}
```

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 15]
```diff
@@ -296,21 +331,24 @@ export class DeploymentManager {
     });
   };
 
-  public setPrinter = (printer: IStackProgressPrinter) => {
-    this.printer = printer;
+  public resetPrinter = (): void => {
+    if (this.printer) {
+      this.printer = null;
+    }
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:337]: This seems to be not very useful. We swap reference anyway below.

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 16]
```diff
@@ -396,13 +434,18 @@ export class DeploymentManager {
     await this.waitForActiveTables(stackParams.tableNames);
   };
 
+  private printLogs = (): void => {
+    this.printer.print();
+  }
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:439]: does it make sense to have alias to a method just to spare `.` ? (if we have more places like this please reconsider, less code is better).

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 17]
```diff
@@ -396,13 +434,18 @@ export class DeploymentManager {
...
+
+  private addPrinterEvent = (event: aws.CloudFormation.StackEvent): void => {
+    this.printer.addActivity(event);
+  }
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:443]: same here.

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts [hunk 18]
```diff
@@ -1,23 +1,123 @@
...
+  private progressBars: MultiProgressBar;
+  private eventMap: EventMap;
+  private categoriesPrinted: string[] = [];
+
+  constructor(eventMap: EventMap) {
+    // Initialize the progress bar and event map.
+    this.progressBars = new MultiProgressBar({
+      progressBarFormatter: createProgressBarFormatter,
+      itemFormatter: createItemFormatter,
+      loneWolf: false,
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts:24]: what's lone wolf?

File: packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+export const CFN_SUCCESS_STATUS = [
+  'UPDATE_COMPLETE', 'CREATE_COMPLETE', 'DELETE_COMPLETE',
+  'DELETE_SKIPPED', 'UPDATE_ROLLBACK_COMPLETE', 'ROLLBACK_COMPLETE',
+];
+export const CNF_ERROR_STATUS = [
+  'CREATE_FAILED', 'DELETE_FAILED', 'UPDATE_FAILED',
+  'UPDATE_ROLLBACK_FAILED', 'ROLLBACK_FAILED',
+];
+
+type ItemPayload = {
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts:14]: What is item?

File: packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+  if (CFN_SUCCESS_STATUS.includes(payload.ResourceStatus)) {
+    color = 'green';
+  }
+  if (CNF_ERROR_STATUS.includes(payload.ResourceStatus)) {
+    color = 'red';
+  }
+  return { renderString, color };
+};
+
+type ProgressPayload = {
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts:49]: payload or key? or just Progress? how do we define progress?

File: packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+  progressName: string,
+  envName: string
+}
+
+/**
+ * Event map type
+ */
+export type EventMap = {
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts:57]: what's a reason to have another type with same name? do properties differ?
Should both `EventMap` present in this PR get more distinct names ?

File: packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+  projectName: string,
+  rootResources: {key: string, category: string}[],
+  eventToCategories: Map<string, string>,
+  categories: {name: string, size: number}[]
+}
+
+/**
+ * Custom progress bar formatter
+ */
+export const createProgressBarFormatter = (
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts:69]: do we export this outside of package?

File: packages/amplify-provider-awscloudformation/src/aws-utils/aws-cfn-progress-formatter.ts (hunk 6 above)
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/src/aws-utils/aws-cfn-progress-formatter.ts:13]: No, it's only consumed in the aws-sdk CloudFormation helper utility in `./aws-cfn.js`

File: packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts (hunk 19 above)
Comment (awsluja) [reply packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts:14]: I found this is a duplicate of the other ItemPayload; and should probably be imported from the 'amplify-prompts' package instead of being re-declared. This code used to live inside of aws-utils/aws-cfn-progress-formatter.ts like this screenshot:

image

File: packages/amplify-prompts/src/progressbars/progressbar.ts [hunk 23]
```diff
@@ -38,7 +38,7 @@ export type ItemPayload = {
     LogicalResourceId: string,
     ResourceType: string,
     ResourceStatus: string,
-    Timestamp: string,
```
[the remark this answers is not in the document]
Comment (awsluja) [reply packages/amplify-prompts/src/progressbars/progressbar.ts:41]: I believe its only use is to describe what the data is sent to be printed in the console:

for example:
deploying...
| resourceId | type | "DEPLOYMENT COMPLETE" | timestamp
| resourceId | type | "DEPLOYMENT COMPLETE" | timestamp
| resourceId | type | "CREATING_IN_PROGRESS" | timestamp

This is internal to the CLI, and only being used in one other package: amplify-provider-awscloudformation , which periodically provides status updates to the printer

I am curious to know more about how we are detecting version bump changes; since we export/import between packages, this may be an issue if packages are dependent on exported types. For example, the data team

File: packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts (hunk 19 above)
Comment (sobolk) [reply packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts:14]: yeah, duplication isn't great (unless we have good reason like not exposing public api surface unnecessarily).
However I'm hoping to prompt discussion about what those properties really represent. `Item` is vague term, not far away from `object`.

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts (hunk 8 above)
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:88]: No

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts (hunk 12 above)
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:204]: I want to keep this PR scoped down to changing the terminal output.

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts (hunk 18 above)
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts:24]: https://github.com/aws-amplify/amplify-cli/blob/dev/packages/amplify-prompts/src/progressbars/progressbar.ts#L10

File: packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts (hunk 22 above)
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts:58]: No

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts (hunk 12 above)
Comment (sobolk) [reply packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:204]: @jhockett yup. How do we keep track of ideas like this?

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts (hunk 14 above)
Comment (sobolk) [reply packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:262]: So, should spinner be intelligent and detect if it has bars or not?
Would we ever have bars without spinner?
I.e. would this be better design?
```
Bars bars = new Bars();
Spinner spinnerWithBars = new Spinner(bars);
try {
 spinnerWithBars.start()
 } finally {
 spinnerWithBars.stop()
 }
 // and then
 Spinner spinnerWithoutBars = new Spinner()
  try {
 spinnerWithoutBars.start()
 } finally {
 spinnerWithoutBars.stop()
 }
```

If this makes sense but is too big feel free to make it backlog item.

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts (hunk 18 above)
Comment (sobolk) [reply packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts:24]: That's surprising.
Why do we pass that flag to multi progress bar then? Isn't multi bar an aggregate of non-lone-wolfs by definition?
Why bars can't detect this themselves through child-parent relationship? (assuming that they form a tree).

File: packages/amplify-prompts/src/progressbars/spinner.ts [hunk 24]
```diff
@@ -3,64 +3,84 @@
  * and use a re writable block instead.
  */
 
-import { AmplifyTerminal as Terminal, StringObj } from './terminal';
+import { AmplifyTerminal, AmplifyTerminal as Terminal, StringObj } from './terminal';
```
Comment (sobolk) [inline packages/amplify-prompts/src/progressbars/spinner.ts:6]: Any chance we can have better name for `StringObj` ?

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 25]
```diff
@@ -1,37 +1,52 @@
...
 
 interface DeploymentManagerOptions {
   throttleDelay?: number;
   eventPollingDelay?: number;
   userAgent?: string;
 }
 
+type SpinnerMessageMap = {
+  [state: string]: { machine: string; message: string };
+};
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:34]:
```suggestion
type SpinnerMessageByState = {
  [state: string]: { machine: string; message: string };
};
```

**Review by sobolk — COMMENTED** [review]
How do we test this change?

File: packages/amplify-prompts/src/progressbars/progressbar.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-prompts/src/progressbars/progressbar.ts:41]: @sobolk I don't think we need to do so, the type is less restrictive.

@awsluja
Lerna looks at the conventional commits to determine which packages were changed and the type of change determined by the git commit message prefix (`feat:` is minor, `chore:` is a patch). Lerna bumps all relevant packages in the dependency tree during the release process.

File: packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts (hunk 21 above)
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts:46]: I agree it's too vague a name. I was going to replace instances of this, but it would cause this PR to balloon in size. I will add this as a follow-up item to refactor the EventMap as it's own PR.

File: packages/amplify-prompts/src/progressbars/progressbar.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (sobolk) [reply packages/amplify-prompts/src/progressbars/progressbar.ts:41]: @jhockett on assignment yes we are more permissive, but on read we become ambiguous.

Would this work after change or not?

```
function foo(timestamp: string) {}

function bar(item: ItemPayload) {
  foo(item.Timestamp); // does this take date?
}
```

[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-prompts/src/progressbars/progressbar.ts]: Great catch, I missed that. I removed the need to change `Timestamp` from a `string` in this exported type.

File: packages/amplify-prompts/src/progressbars/multibar.ts [hunk 26]
```diff
@@ -14,21 +14,21 @@ import {
...
  */
 export class MultiProgressBar {
-    private count : number;
-    private terminal : Terminal;
-    private options : BarOptions;
-    private bars : {name: string, bar: Bar}[];
+    private count: number;
+    private terminal: Terminal;
+    private options: BarOptions;
+    private bars: {name: string, bar: Bar}[];
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-prompts/src/progressbars/multibar.ts]: https://github.com/aws-amplify/amplify-cli/pull/11051/files?w=1

File: packages/amplify-prompts/src/progressbars/spinner.ts (hunk 4 above)
Comment (jhockett) [reply packages/amplify-prompts/src/progressbars/spinner.ts:52]: I don't have the full context behind this change

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts (hunk 18 above)
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts:24]: It's due to a break down in abstractions. When a progressbar is part of a multibar, the multibar handles the rendering. Added a backlog item to refactor this into a cleaner abstraction.

File: packages/amplify-prompts/src/progressbars/spinner.ts (hunk 5 above)
Comment (jhockett) [reply packages/amplify-prompts/src/progressbars/spinner.ts:84]: I think that might make sense, I've added this as a backlog item to refactor.

I'm not aware of such a concept in TS.

Preventing bad usage is up to the caller. This is similar to how we use `ora`.

**Review by sobolk — DISMISSED** [review]

File: packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts (hunk 19 above)
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/src/utils/progress-bar-helpers.ts]: I fixed the import and added a backlog task to refactor ItemPayload to a more descriptive name.

**Comment by awsluja** [comment]
Seeing some issues on the e2e run with the 'schema-iterative-rollback' tests, looks like we're expecting to see "Deploying (X of Y)" during iterative deployments. Looking at the artifacts, I didn't see an iterative deployment performed, so this might also be a regression, or there are other issues earlier in the test

File: packages/amplify-prompts/src/progressbars/multibar.ts [hunk 27]
```diff
@@ -14,21 +14,21 @@ import {
...
     private lastDrawnTime: number;
     isActive: boolean;
     private refreshRate: number;
-    private frameCount : number;
-    private frames : string[];
+    private frameCount: number;
+    private frames: string[];
     private timer!: ReturnType<typeof setTimeout>;
-    private prefixText : string;
+    private prefixText: string; // Header text goes in front of an animated spinner
```
Comment (sobolk) [inline packages/amplify-prompts/src/progressbars/multibar.ts:27]: should we call the field `header` then ?

File: packages/amplify-prompts/src/progressbars/multibar.ts [hunk 28]
```diff
@@ -14,21 +14,21 @@ import {
...
     private updated: boolean;
-    private lastDrawnStrings : StringObj[];
+    private lastDrawnStrings: TerminalLine[];
```
Comment (sobolk) [inline packages/amplify-prompts/src/progressbars/multibar.ts:29]: consider renaming field to `trailingLines` or `suffix` or `footer`.

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts [hunk 29]
```diff
@@ -59,21 +59,29 @@ export class StackProgressPrinter implements IStackProgressPrinter {
...
       }
       this.progressBars.create(progressBarsConfigs);
     }
   };
 
+  updateIndexInHeader = (currentIndex: number, totalIndices: number): void => {
+    this.progressBars.updatePrefixText(
+      `Deploying iterative update ${currentIndex} of ${totalIndices} into ${this.eventMap.envName} environment. This will take a few minutes.`,
+    );
+  }
```
Comment (sobolk) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts:83]: Should this be `updateRelativeProgressCounter` or something like this?
Or perhaps should be this be `onProgress(completed, total)` ? The fact that this updates header is implementation details of this component.

File: packages/amplify-prompts/src/progressbars/multibar.ts (hunk 27 above)
Comment (jhockett) [reply packages/amplify-prompts/src/progressbars/multibar.ts:27]: Added to backlog refactor task

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts (hunk 29 above)
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts:83]: Great catch, the class shouldn't leak implementation details. Added to backlog refactor task.

**Review by sobolk — DISMISSED** [review]

File: packages/amplify-prompts/src/progressbars/progressbar.ts [hunk 30]
```diff
@@ -69,7 +69,9 @@ export class ProgressBar {
     barSize: number;
 
     constructor(options: BarOptions) {
-      this.terminal = new Terminal();
+      if (options.loneWolf) {
+        this.terminal = new Terminal();
+      }
```
Comment (evcodes) [inline packages/amplify-prompts/src/progressbars/progressbar.ts:74]: what does lonewolf mean here? Does it mean that there is no terminal such that the progress bar can be attached to?

File: packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts [hunk 31]
```diff
@@ -1,37 +1,52 @@
...
+const rollbackSpinnerMessages: SpinnerMessageByState = {
+  idle: { machine: 'rollback', message: `Starting rollback.` },
+  preRollback: { machine: 'rollback', message: `Waiting for previous deployment to finish.` },
+  'rollback.waitForTablesToBeReady': { machine: 'rollback', message: `Waiting for DynamoDB indices to be ready.` },
+};
+
 /**
- *
+ * Deployment error class
+ * TODO conform to new error handling mechanism
```
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/src/iterative-deployment/deployment-manager.ts:49]: TODO before merging this PR?

File: packages/amplify-prompts/src/progressbars/progressbar.ts (hunk 30 above)
Comment (jhockett) [reply packages/amplify-prompts/src/progressbars/progressbar.ts:74]: Discussed above in a resolved comment, but it indicates if the progressbar is part of a multibar or not. If the progress bar has `loneWolf` set to true, it's a standalone will handle the rendering. If false, the parent class will take care of rendering the various progress bars. The confusing naming and leaky abstraction will be addressed in a future refactor PR.

**Review by awsluja — APPROVED** [review]
LGTM , nice work fixing e2e, I have a comment regarding test coverage but I don't think its blocking

**Review by edwardfoyle — APPROVED** [review]

**Review by sobolk — APPROVED** [review]

**Review by awsluja — APPROVED** [review]
LGTM
